client/src/main/java/org/asynchttpclient/cookie/ThreadSafeCookieStore.java

```diff
@@ -145,45 +161,76 @@ private void add(String requestDomain, String requestPath, Cookie cookie) {
     String keyDomain = pair.getKey();
     boolean hostOnly = pair.getValue();
     String keyPath = cookiePath(cookie.path(), requestPath);
-    CookieKey key = new CookieKey(cookie.name().toLowerCase(), keyDomain, keyPath);
+    CookieKey key = new CookieKey(cookie.name().toLowerCase(), keyPath);
 
     if (hasCookieExpired(cookie, 0))
-      cookieJar.remove(key);
-    else
-      cookieJar.put(key, new StoredCookie(cookie, hostOnly, cookie.maxAge() != Cookie.UNDEFINED_MAX_AGE));
+      cookieJar.getOrDefault(keyDomain, Collections.emptyMap()).remove(key);
+    else {
+      cookieJar.putIfAbsent(keyDomain, new ConcurrentHashMap<>());
+      cookieJar.get(keyDomain).put(key, new StoredCookie(cookie, hostOnly, cookie.maxAge() != Cookie.UNDEFINED_MAX_AGE));
+    }
   }
 
   private List<Cookie> get(String domain, String path, boolean secure) {
 
-    final boolean[] removeExpired = {false};
+    final List<Cookie> results = new ArrayList<>();
+    boolean exactDomainMatch = true;
+    String topLevelSubdomain = domain;
 
-    List<Cookie> result = cookieJar.entrySet().stream().filter(pair -> {
+    while (MiscUtils.isNonEmpty(topLevelSubdomain)) {
+      results.addAll(getStoredCookies(topLevelSubdomain, path, secure, exactDomainMatch));
+      exactDomainMatch = false;
+      topLevelSubdomain = DomainUtils.getTopLevelSubdomain(topLevelSubdomain);
```

Comment:
Taking advantage that:

1. Domain matches exactly, in which case it does not matter if its hostOnly domain or not
2. Request domain is subdomain of cookie domain.